packages/frames.js/src/core/middleware/parallelizeMiddleware.ts

```diff
@@ -0,0 +1,36 @@
+import type { FramesContextFromMiddlewares, FramesMiddleware } from "../types";
+
+/**
+ * Parallelizes provided middlewares. Each middleware should be isolated and not depend on some context value that could be provided by some other middleware that can potentially run in parallel.
+ *
+ * Also next function passed to each middleware does nothing except modifying the context.
+ */
+export function parallelizeMiddleware<
+  TFramesMiddleware extends FramesMiddleware<any>[],
+>(
+  ...middlewares: TFramesMiddleware
+): FramesMiddleware<FramesContextFromMiddlewares<TFramesMiddleware>> {
+  if (middlewares.length === 0) {
+    throw new Error("No middlewares provided");
+  }
+
+  if (middlewares.length === 1) {
+    return middlewares[0]!;
+  }
+
+  return async (context, next) => {
+    let ctx = context;
+
+    await Promise.all(
+      middlewares.map((middleware) =>
+        // @ts-expect-error this is not 100% correct because next should return a valid result, but since parallel middleware cannot call next middleware in chain, it's fine
+        middleware(context, (newCtx) => {
+          ctx = { ...ctx, ...newCtx };
+        })
+      )
+    );
```

Comment:
are we sure this works and there are no race conditions that could cause some returned context to be skipped? would it not be better to collect all the new contexts and reduce them into an aggregate context?

Reply:
there can't be a race condition since this is not truly parallel but I will change the logic to make sure that `ctx` inside the function is not a copy therefore it could overwrite stuff.

Reply:
@stephancill I renamed the function and also changed it to aggregate new contexts and then merge them at the end.

Reply:
thank you!